Linting and Formatting Upgrade

connorwinston opened the pull request:
See https://github.com/directus/directus/pull/18892#issuecomment-1607633804 for updated description.

---

- Enabled linting of files in `/docs/.vitepress` and `/docs/.typedocs`
- Enabled linting of `.md` mono-repo wide (Added `eslint-plugin-md`)
- Fixed formatting of  all`.md` files mono-repo wide
- Added `prettier-ignore` comments to invalid docs code blocks. 
  - You can now run `pnpm format` mono-repo wide without fear of messing up docs comments.
- Manually checked all docs pages to ensure no formatting got messed up in this upgrade. @phazonoverload 😀

Closes #18876

connorwinston committed:
add docs to eslint

connorwinston committed:
update prettier ignore

connorwinston committed:
fix vitepress linting

connorwinston committed:
eslint ignore fixes

connorwinston committed:
prettier run

connorwinston committed:
update prettier ignore

connorwinston committed:
fix formatting

connorwinston committed:
enable linting of markdown files

connorwinston committed:
revert format command change

connorwinston committed:
fix irregular whitespace

connorwinston committed:
update dictionary

connorwinston committed:
(Changelog) Create four-boxes-shake.md

paescuj commented:
Thanks @ConnorSimply, this is great ❤️ I'm going to adjust a few things in this PR and will mark it as "Ready for review" again afterwards.

rijkvanzanten commented:
> Thanks @ConnorSimply, this is great ❤️ I'm going to adjust a few things in this PR and will mark it as "Ready for review" again afterwards.

What are you gonna change? Lets just put the needed changes in a review request, otherwise nobody knows what we're waiting for 😅

paescuj committed:
Rework ESLint / Prettier setup

- Disable js/ts/vue files for Prettier to ensure linting/formatting is
  only happening via ESLint
- Rework formatting of code blocks in md files
  - Disable formatting of code blocks in md files under '/docs' by Prettier
  - Instead use "eslint-plugin-markdown" to format & __lint__ js*/ts*/vue such code blocks
  - Replace unmaintained "eslint-plugin-md" plugin by official "eslint-plugin-markdown" plugin
  - I'll check whether we can use this to format other code blocks
    (json, html, ...) as well
- Restructure, clean-up and apply some fixes to the ESLint config
  (Note: Not ready for flat config yet since not supported by
  vscode-eslint)
- Enable cache for ESLint / Prettier in scripts
- Clean-up ignore file
  - Explicit folder declaration (.../)
  - Don't ignore all 'extensions' folders in ESLint (only
    '/api/extensions/')
  - Enable formatting in '/.github' folder

paescuj committed:
Fix all formatting issues with Prettier

paescuj committed:
Update md files under /docs/.typedocs

paescuj committed:
Fix lint issues in vue/js files

ConnorSimply commented:
@paescuj Is this ready to be reviewed and merged?

paescuj committed:
ESLint / Prettier config revision v2

Enable Prettier for md code blocks, but only as warnings since it can
get into the way with Vitepress md extensions like '[!code ...]'
comments

paescuj committed:
Remove prettier-ignore comments

paescuj committed:
Merge branch 'main' into connorsimply/eng-1266-enable-linting-for-the-docs-config-and-custom-theme

paescuj committed:
Make spellchecker happy

paescuj committed:
Remove changeset

paescuj committed:
Revert lint setup for code blocks

There are many cases in the docs where linting / formatting of code
blocks doesn't make
sense:
- Code block is only an excerpt - linter fails
- Code block contains special comments (e.g. markdown extensions) which
  needs to remain at the same place - formatting would break it
- ...

paescuj committed:
Apply lint issues / formatting from temp lint setup

paescuj committed:
Run formatter

paescuj committed:
Merge branch 'main' into connorsimply/eng-1266-enable-linting-for-the-docs-config-and-custom-theme

paescuj committed:
Fix merge failure

paescuj committed:
Simplify & modernize ESLint / Prettier setup

No longer run Prettier via ESLint. Nowadays, this is the recommended
setup. There's no real need to run it this way, it's just an additional
layer.

Add VS Code settings to make the work with the new setup easier.

paescuj committed:
Remove unused eslint disable directives

paescuj committed:
Make editorconfig more useful

paescuj committed:
Fix formatting issues reported by editorconfig

paescuj committed:
Format files with Prettier

paescuj committed:
Enable formatting of source translations file

paescuj committed:
Format source translations file

paescuj commented on docs/.vitepress/components/ArticleFeedback.vue:
```diff
@@ -81,7 +83,7 @@ async function handleSubmission(rating?: number) {
 	} catch (err) {
 		error.value = err;
-		console.error(err);
+		// console.error(err);
```
I think we can remove this altogether
```suggestion
```

paescuj committed:
Remove unnecessary console error

paescuj committed:
Remove unnecessary line

paescuj committed:
Only ignore md files under .changeset

paescuj committed:
Add CI reporter for Prettier

paescuj committed:
Merge branch 'main' into connorsimply/eng-1266-enable-linting-for-the-docs-config-and-custom-theme

paescuj committed:
Fail job on wrongly formatted files

paescuj committed:
Fix format

paescuj committed:
Test Prettier action on changed/added file

paescuj committed:
Use simple CI format check for now & no cache

paescuj committed:
Revert "Test Prettier action on changed/added file"

This reverts commit 4f7d8826ad877be066f3a327c37b62672c0251fb.

paescuj committed:
Introduce code blocks check for docs

paescuj committed:
Fix code block issues

paescuj committed:
Ignore auto-generated packages dir

paescuj committed:
Fix comment position

paescuj commented:
Sooo finally my owed explanation 🎉 

Updating the ESLint/Prettier setup is something I have been meaning to do for a while and I wanted to continue on Connor's initiative (thanks again ❤️) to make it more usable for the docs.
The final setup now consists of the following:

- No longer run Prettier via ESLint. Nowadays, this is the recommended way.
  - https://prettier.io/docs/en/integrating-with-linters.html#notes sums up the motivation behind it pretty well.
  - Makes the setup more flexible in general and allows us to easily run Prettier over the whole codebase be it via the CLI / Editor / CI.

- Enforce formatting via GitHub workflow.

- Disable formatting of code blocks in the docs.
  - This has the potential to break too much things, like `[!code ..]` comments.
  - Safer to turn it off completely instead of for individual blocks. Adding `prettier-ignore` comments to prevent formatting can be easily forgotten and at the same time it's hard to spot when things have been messed up.
 
- Clean-up as well as fixing some issues in the ESLint configuration.
  - For example some rules weren't applied to Vue files before (like `prefer-const`).
  - Clearer separation of the rules.
  - Removed `jest` plugin:  Wasn't configured correctly anyway and will be superfluous in the future.

- Add linting & formatting configuration for VS Code.
  - We can say pretty confidently that VS Code is the most used editor for devs working on our codebase, so I think it makes sense to add a configuration for it.
  - Since we require linting & formatting anyway, it makes sense to enable this by default for the editor.

- Enable local caching for both ESLint and Prettier to speed up things.

- Make editorconfig more useful.
  - It doesn't make sense to enforce tabs for _all_ files.
  - Now files can also be checked, for example via `editorconfig-checker`. Helped me to find issues like https://github.com/directus/directus/pull/18890/commits/337227d29162f64e55beeb7cf6d4838776bebbf8.

- Clean-up of ignore files.
  - Still enabling linting of files in `/docs/.vitepress` and `/docs/.typedocs` and additionally enable formatting for `en-US.yaml` file. 

---

Linting and formatting of code blocks in Markdown files is interesting but we cannot have it enabled by default.
It interferes too much, can mess things up and gives too many false negatives.
On the other side it proved to be helpful in finding quite some errors (syntax issues, wrong code block types, ...) and to improve formatting. For this reason I've left the possibility to check it manually:
```
pnpm --filter docs check-code-blocks
```

---

All lint & format issues have been fixed and double-checked.

paescuj committed:
Also lint `/app/.storybook`

paescuj committed:
Merge branch 'main' into connorsimply/eng-1266-enable-linting-for-the-docs-config-and-custom-theme

paescuj committed:
Reformat modified files

rijkvanzanten committed:
Merge branch 'main' into connorsimply/eng-1266-enable-linting-for-the-docs-config-and-custom-theme

rijkvanzanten committed:
Merge branch 'main' into connorsimply/eng-1266-enable-linting-for-the-docs-config-and-custom-theme
